src/hotspot/share/gc/g1/g1ConcurrentRefineThreadsNeeded.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "gc/g1/g1Analytics.hpp"
+#include "gc/g1/g1ConcurrentRefineThreadsNeeded.hpp"
+#include "gc/g1/heapRegion.hpp"
+#include "gc/g1/g1Policy.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include <math.h>
+
+G1ConcurrentRefineThreadsNeeded::G1ConcurrentRefineThreadsNeeded(G1Policy* policy,
+                                                                 double update_period_ms) :
+  _policy(policy),
+  _update_period_ms(update_period_ms),
+  _predicted_time_until_next_gc_ms(0.0),
+  _predicted_cards_at_next_gc(0),
+  _threads_needed(0)
+{}
+
+// Estimate how many concurrent refinement threads we need to run to achieve
+// the target number of card by the time the next GC happens.  There are
+// several additional desirements we'd like to achieve while meeting that
+// goal.
+//
+// 1. Minimize the number of refinement threads running at once.
+//
+// 2. Minimize the number of activations and deactivations for the
+// refinement threads that run.
+//
+// 3. Delay performing refinement work.  Having more dirty cards waiting to
+// be refined can be beneficial, as further writes to the same card don't
+// create more work.
+void G1ConcurrentRefineThreadsNeeded::update(uint active_threads,
+                                             size_t available_bytes,
+                                             size_t num_cards,
+                                             size_t target_num_cards) {
+  const G1Analytics* analytics = _policy->analytics();
+
+  // Estimate time until next GC, based on remaining bytes available for
+  // allocation and the allocation rate.
+  double alloc_region_rate = analytics->predict_alloc_rate_ms();
+  double alloc_bytes_rate = alloc_region_rate * HeapRegion::GrainBytes;
+  if (alloc_bytes_rate == 0.0) {
+    _predicted_time_until_next_gc_ms = 0.0;
```

Comment:
A zero value for the prediction indicates that we don't have a valid
prediction, because we're starting up the VM and don't have data yet. That
puts us in a place where we can't really do much with the predictive model.
Later clauses (no refinement needed, or time is short) also "handle" this
case, though maybe there should be some commentary about that.

Reply:
> A zero value for the prediction indicates that we don't have a valid
prediction

Why not? It's still possible that the alloc-rate is zero after start-up; I mean alloc-rate is up to applications.

On a related note, there's special treatment for too-close upcoming GC pause later on, `if (_predicted_time_until_next_gc_ms > _update_period_ms) {`. Shouldn't there be sth similar for too-far upcoming GC pause? IOW, `incoming_rate * _predicted_time_until_next_gc_ms;` would be unreliable for farther prediction, right?

Reply:
> Why not? It's still possible that the alloc-rate is zero after start-up; I mean alloc-rate is up to applications.

Allocation rate is the rate of allocation between GCs, to have a GC, you (almost) need non-zero allocation rate (not with periodic gcs).

Reply:
I've added some comments regarding the alloc_bytes_rate == 0 case.

I don't think we need to do anything special for predicted GC times being far
away, because we'll be periodically re-evaluating.